bundles/org.openhab.binding.heliosventilation/src/main/java/org/openhab/binding/heliosventilation/internal/HeliosVentilationHandler.java

```diff
@@ -0,0 +1,454 @@
+                poll(v);
+            }
+        });
+    }
+
+    private void disconnect() {
+        if (thing.getStatus() != ThingStatus.REMOVING) {
+            updateStatus(ThingStatus.OFFLINE);
+        }
+        synchronized (this) {
+            try {
+                if (inputStream != null) {
+                    inputStream.close();
+                }
+            } catch (IOException e) {
+                // ignore the exception on close
+                inputStream = null;
+            }
+            try {
+                if (outputStream != null) {
+                    outputStream.close();
+                }
+            } catch (IOException e) {
+                // ignore the exception on close
+                outputStream = null;
+            }
+
+            SerialPort serial = serialPort;
+            if (serial != null) {
+                serial.close();
+            }
+            serialPort = null;
+        }
+
+    }
+
+    private void poll(HeliosVentilationDataPoint v) {
+        byte txFrame[] = { 0x01, BUSMEMBER_ME, BUSMEMBER_MAINBOARD, 0x00, v.address(), 0x00 };
+        txFrame[5] = (byte) checksum(txFrame);
+
+        tx(txFrame);
+    }
+
+    /*
+     * transmit a frame
+     */
+    private void tx(byte[] txFrame) {
+        try {
+            OutputStream out = outputStream;
+            if (out != null) {
+                if (logger.isTraceEnabled()) {
+                    logger.trace("HeliosVentilation: Write to serial port: {}",
+                            String.format("%02x %02x %02x %02x", txFrame[1], txFrame[2], txFrame[3], txFrame[4]));
+                }
+
+                out.write(txFrame);
+                out.flush();
+                // after each frame we have to wait.
+                // 30 ms is taken from what we roughly see the original remote control is doing
+                Thread.sleep(30);
+            }
+        } catch (IOException e) {
+            // in case we cannot write the connection is somehow broken, let's officially disconnect
+            disconnect();
+            connect();
+        } catch (InterruptedException e) {
+            // ignore if we got interrupted
+        }
+    }
+
+    /**
+     * Check connection status
+     *
+     * @return true if currently connected
+     */
+    private boolean isConnected() {
+        return serialPort != null && inputStream != null && outputStream != null;
+    }
+
+    @Override
+    public synchronized void serialEvent(SerialPortEvent event) {
+        switch (event.getEventType()) {
+            case SerialPortEvent.DATA_AVAILABLE:
+                // we get here if data has been received
+
+                try {
+                    // Wait roughly a frame length to increase the likelihood that the complete frame is already
+                    // buffered. This improves the robustness for RS485/USB converters which sometimes duplicate bytes
+                    // otherwise
+                    Thread.sleep(8);
```

Comment:
why do you mean "increase the likelihood"? Is it possible to implement this in a foolproof manner?

Reply:
I recently changed that already compared to the initial (abandoned) PR from 5 to 8, which was the change from "improving" to surly waiting until the frame is fully buffered. Will update the comment accordingly.